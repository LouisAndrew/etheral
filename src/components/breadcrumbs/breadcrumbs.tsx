import React from 'react';
import { Link } from '@reach/router';

import { Text, Flex } from 'rebass';
import { InlineIcon } from '@iconify/react';
import arrowRightSLine from '@iconify/icons-ri/arrow-right-s-line';

import { nameToSlug } from 'helper/const';

/**
 * @param location: Location of breadcrumb's parent component.
 */
type Props = {
    location: string;
    append?: boolean;
    appendText?: string;
};

const Breadcrumbs: React.FC<Props> = ({ location, append, appendText }) => {
    const locationSplitted = append
        ? [...location.split('/'), appendText || '']
        : location.split('/');

    return (
        <Flex sx={{ a: { textDecoration: 'none' } }}>
            {/* base home link. */}
            <Link to="/">
                <Text
                    variant="body"
                    fontWeight="regular"
                    ml={[1]}
                    sx={{
                        svg: {
                            ml: [1],
                            transform: 'translate(0, 4px)',
                        },
                    }}
                >
                    HOME
                    <InlineIcon icon={arrowRightSLine} />
                </Text>
            </Link>
            {/* rendered links from location props */}
            {locationSplitted.map((loc, i) => {
                if (append && i === locationSplitted.length - 1) {
                    return (
                        <Text variant="body" fontWeight="bold">
                            {loc}
                        </Text>
                    );
                } else {
                    const href = locationSplitted
                        .slice(0, i + 1)
                        .map(link => nameToSlug(link)) // reformat all location member to have its "slug attribute"
                        .join('/'); // reconfigure the link for EACH location string.

                    return (
                        <Link to={href} key={`breadcrumb-link-${href}`}>
                            <Text
                                variant="body"
                                fontWeight={
                                    i === locationSplitted.length - 1
                                        ? 'bold'
                                        : 'regular'
                                }
                                ml={[1]}
                                sx={{
                                    svg: {
                                        ml: [1],
                                        transform: 'translate(0, 4px)',
                                    },
                                }}
                            >
                                {loc.toUpperCase()}
                                {i !== locationSplitted.length - 1 && (
                                    <InlineIcon icon={arrowRightSLine} />
                                )}
                            </Text>
                        </Link>
                    );
                }
            })}
        </Flex>
    );
};

export { Breadcrumbs };
