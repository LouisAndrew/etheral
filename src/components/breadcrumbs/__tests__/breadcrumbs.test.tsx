import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';

import { Breadcrumbs } from '../breadcrumbs';

describe('Breadcrumbs', () => {
    const mockAppendText = 'SHOP ALL';

    const Element = <Breadcrumbs location="shop/collection1/fall" />;
    const ElmentWithAppend = (
        <Breadcrumbs
            location="shop"
            append={true}
            appendText={mockAppendText}
        />
    );

    afterEach(cleanup);

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(Element, div);
    });

    // // hard-coded the test for simplicity sake
    // it('should render links from the location correctly', () => {
    //     const { getByText } = render(Element);

    //     const Collection1El = getByText('COLLECTION1');

    //     expect(Collection1El.getAttribute('href')).toBe('shop/collection1');
    // });

    it('Should render the appended text', () => {
        const { queryByText } = render(ElmentWithAppend);

        const AppendedText = queryByText(mockAppendText);

        if (AppendedText) {
            expect(AppendedText).toBeInTheDocument();
        } else {
            fail();
        }
    });

    it('matches snapshot', () => {
        const run = true;

        if (run) {
            const tree = renderer.create(Element).toJSON();
            expect(tree).toMatchSnapshot();
        }
    });
});
