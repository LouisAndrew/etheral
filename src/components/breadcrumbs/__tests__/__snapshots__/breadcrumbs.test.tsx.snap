// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs matches snapshot 1`] = `
<div
  className="css-tcs43g"
>
  <a
    aria-current="page"
    href="/"
    onClick={[Function]}
  >
    <div
      className="css-10v0p1r"
    >
      HOME
      <svg
        dangerouslySetInnerHTML={
          Object {
            "__html": "<path d=\\"M13.172 12l-4.95-4.95l1.414-1.414L16 12l-6.364 6.364l-1.414-1.414z\\" fill=\\"currentColor\\"/>",
          }
        }
        focusable={false}
        height="1em"
        preserveAspectRatio="xMidYMid meet"
        style={
          Object {
            "transform": "rotate(360deg)",
            "verticalAlign": "-0.125em",
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      />
    </div>
  </a>
  <a
    href="/shop"
    onClick={[Function]}
  >
    <div
      className="css-10v0p1r"
    >
      SHOP
      <svg
        dangerouslySetInnerHTML={
          Object {
            "__html": "<path d=\\"M13.172 12l-4.95-4.95l1.414-1.414L16 12l-6.364 6.364l-1.414-1.414z\\" fill=\\"currentColor\\"/>",
          }
        }
        focusable={false}
        height="1em"
        preserveAspectRatio="xMidYMid meet"
        style={
          Object {
            "transform": "rotate(360deg)",
            "verticalAlign": "-0.125em",
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      />
    </div>
  </a>
  <a
    href="/shop/collection1"
    onClick={[Function]}
  >
    <div
      className="css-10v0p1r"
    >
      COLLECTION1
      <svg
        dangerouslySetInnerHTML={
          Object {
            "__html": "<path d=\\"M13.172 12l-4.95-4.95l1.414-1.414L16 12l-6.364 6.364l-1.414-1.414z\\" fill=\\"currentColor\\"/>",
          }
        }
        focusable={false}
        height="1em"
        preserveAspectRatio="xMidYMid meet"
        style={
          Object {
            "transform": "rotate(360deg)",
            "verticalAlign": "-0.125em",
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      />
    </div>
  </a>
  <a
    href="/shop/collection1/fall"
    onClick={[Function]}
  >
    <div
      className="css-16mev7"
    >
      FALL
    </div>
  </a>
</div>
`;
